File: Ahorn/entities/TesseractBoss.jl
module TesseractBoss

using ..Ahorn, Maple

@mapdef Entity "MaggyHelper/TesseractBoss" TesseractBoss(x::Integer, y::Integer, health::Integer=1350, maxHealth::Integer=1350)

const placements = Ahorn.PlacementDict(
    "tesseract_boss" => Ahorn.EntityPlacement(TesseractBoss)
)

function Ahorn.selection(entity::TesseractBoss)
    return Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::TesseractBoss, room::Maple.Room)
    Ahorn.drawRectangle(ctx, Ahorn.Rectangle(entity.x - 8, entity.y - 8, 16, 16), (0.5, 0.5, 0.5, 0.5), (0.0, 0.0, 0.0, 1.0))
end

end

File: Loenn/entities/tesseract_boss.lua
local drawableSprite = require("structs.drawable_sprite")
local utils = require("utils")

local tesseractBoss = {}

tesseractBoss.name = "MaggyHelper/TesseractBoss"
tesseractBoss.depth = -8500
tesseractBoss.justification = {0.5, 0.5}

tesseractBoss.placements = {
    {
        name = "tesseract_boss",
        data = {
            health = 1350,
            maxHealth = 1350,
    }
    }
}

tesseractBoss.fieldInformation = {
    health = {
        fieldType = "integer",
        minimumValue = 1
    },
    maxHealth = {
        fieldType = "integer",
        minimumValue = 1
    }
}

tesseractBoss.fieldOrder = {
    "x", "y",
    "health",
    "maxHealth"
}

function tesseractBoss.sprite(room, entity)
    local textures = {
        "objects/tesseract_temple/dashButton00",
        "objects/tesseract_temple/dashButtonMirror00",
        "characters/Kglobal::Player/sitDown00"
    }

    for _, texture in ipairs(textures) do
        local ok, sprite = pcall(drawableSprite.fromTexture, texture, entity)
        if ok and sprite then
            sprite:setJustification(0.5, 0.5)
            return {sprite}
        end
    end

    return {}
end

function tesseractBoss.selection(room, entity)
    return utils.rectangle(entity.x - 27, entity.y - 27, 54, 54)
end

return tesseractBoss
